functionary/prompt_template/base_template.py

```diff
@@ -125,6 +127,18 @@ def get_prompt_from_messages(
             str: the prompt for inference/training
         """
 
+        for i in range(len(tools_or_functions)):
+            if "type" in tools_or_functions[i]:
```

Comment:
I think it is better to copy a new variable for tools_or_functions to make sure that original tools_or_functions is intact as this can be used in other locations.

tools = copy.deepcopy(tools_or_functions)
....